src/main/java/ir/mimlang/jmim/lang/interpreter/Interpreter.kt

```diff
@@ -9,566 +9,583 @@ import ir.mimlang.jmim.lang.util.ext.*
-					popContext()
-					return null
-				}
-				
-				return null
-			}
-			
-			is NamedBlockNode -> {
-				when (node.name) {
-					"repeat" -> {
-						breakable {
-							while (true) {
-								continuable {
-									pushContext("infinite repeat block")
-									node.body.nodes.interpret()
-									popContext()
-								} onContinue { popContext() }
-							}
-						} onBreak { popContext() }
-						
-						return null
-					}
-					
-					"run" -> {
-						pushContext("run block")
-						val result = node.body.nodes.interpretAndReturnLast()
-						popContext()
-						
-						return result
-					}
-					
-					"list" -> return node.body.nodes.map(this::interpret)
-					
-					"map" -> {
-						val map = mutableMapOf<String, Any?>()
-						for (i in node.body.nodes.indices step 2) {
-							val key = (interpret(node.body.nodes[i]) as? String)
-								?: throw InterpreterException("map key must be string") at node.body.nodes[i].range
-							val value = node.body.nodes.getOrNull(i + 1)?.let { interpret(it) }
-							
-							map[key] = value
-						}
-						
-						return map
-					}
-					
-					"func" -> return FunctionVariable("lambda${node.range}", listOf(), node.body.nodes)
-					
-					else -> throw InterpreterException("Named block with name '${node.name}' is not known") at node.range
-				}
-			}
-			
-			else -> throw InterpreterException("Unsupported node") at node.range
-		}
-	}
-	
-	private fun pushContext(name: String) {
-		context = Context(context, name)
-	}
-	
-	private fun popContext() {
-		context.parent?.let { context = it }
-	}
-	
-	private fun List<Node>.interpret() = forEach(this@Interpreter::interpret)
-	
-	private fun List<Node>.interpretAndReturnLast(): Any? {
-		isEmpty() then { return null }
-		
-		for (i in 0 until lastIndex) interpret(get(i))
-		return interpret(last())
-	}
-	
-	private fun CaseOrGroupNode.check(
-		operand: ParenthesizedOperationNode,
-		defaultComparator: String?
-	): Boolean {
-		val cond = groups.let orResult@{ orGroup ->
-			orGroup.forEach { andGroup ->
-				val andResult = andGroup.cases.let andResult@{ cases ->
-					cases.forEach { case ->
-						val caseResult = interpret(
-							BinaryOperationNode(
-								operand,
-								case.operator ?: defaultComparator ?: throw InterpreterException("no operator found to check case condition") at case.range,
-								case.operand,
-								case.range
-							)
-						) as Boolean
-						if (!caseResult) return@andResult false
-					}
-					return@andResult true
-				}
-				if (andResult) return@orResult true
-			}
-			return@orResult false
-		}
-		return cond
-	}
+    fun interpret(node: Node): Any? {
+        when (node) {
+            is StatementNode -> return interpret(node.node)
+            is ParenthesizedOperationNode -> return interpret(node.node)
+
+
+            is StringNode -> {
+                if (node.isRaw) return node.string else {
+                    val result = Regex("\\$[a-zA-Z-0-9-_]+").findAll(node.string).toList()
```

Comment:
your regex is wrong here. please see [line 10](https://github.com/MohammadMD1383/J-MiM/blob/f1544281ce7c25f329578c90405370f261b22957/src/main/java/ir/mimlang/jmim/lang/token/Tokens.kt#L10) and [line 11](https://github.com/MohammadMD1383/J-MiM/blob/f1544281ce7c25f329578c90405370f261b22957/src/main/java/ir/mimlang/jmim/lang/token/Tokens.kt#L11) of `Tokens.kt` for more information on how to detect an **identifier**.